### PR TITLE
Add admin keygen experimental CLI command

### PR DESCRIPTION
**Pull request by peterschwarz**
This PR adds the subcommand `admin keygen` to the grid CLI.  This command generates a key for the grid daemon to use when creating scabbard services on splinter circuits and initializing the service with the necessary grid smart contracts.

It is an experimental feature guarded by the feature `"admin-keygen"`.

Additionally, 
- The deb package will create the directory `/etc/grid/keys`, where this key will be stored.

### CODE REVIEW

File: cli/src/actions/admin.rs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+use std::fs::OpenOptions;
+use std::io::prelude::*;
+use std::os::unix::fs::OpenOptionsExt;
+use std::path::PathBuf;
+
+use sawtooth_sdk::signing;
+
+use crate::error::CliError;
+
+const KEY_DIR: &str = "/var/lib/grid";
```
Comment (rbuysse) [inline cli/src/actions/admin.rs]: Should this be `/etc/grid/keys`?

File: cli/src/actions/admin.rs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+const KEY_DIR: &str = "/etc/grid/keys";
+
+pub enum ConflictStrategy {
+    Force,
+    Skip,
+    Error,
+}
+
+pub fn do_keygen(conflict_strategy: ConflictStrategy) -> Result<(), CliError> {
+    let public_key_path: PathBuf = [KEY_DIR, "grid.pub"].iter().collect();
```
Comment (vaporos) [inline cli/src/actions/admin.rs]: I think we should probably call these 'gridd.pub' and 'gridd.priv' so it is clear that these are the ones used by the grid daemon.

File: cli/src/actions/admin.rs (hunk 1 above)
Comment (peterschwarz) [reply cli/src/actions/admin.rs]: Fixed

File: cli/src/actions/admin.rs (hunk 2 above)
Comment (peterschwarz) [reply cli/src/actions/admin.rs]: updated.

File: cli/src/actions/admin.rs [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+            if !public_key_path.exists() && private_key_path.exists() {
+                return Err(CliError::UserError(format!(
+                    "{} already exists without a corresponding public key. \
+                     Rerun with --force to overwrite existing files",
+                    private_key_path.as_path().display(),
+                )));
+            }
+
+            if public_key_path.exists() && private_key_path.exists() {
+                info!("Admin private keys exists; skipping generation.");
```
Comment (rbuysse) [inline cli/src/actions/admin.rs]: Small grammar issue here.

**Comment by rbuysse** [comment]
Can you update this so the default behavior is similar to what happens when using `-v`?

```
# grid admin keygen
#
```
```
# grid -v admin keygen  
2020-01-02 21:19:59,628 INFO  [grid::actions::admin] Writing file: "/etc/grid/keys/gridd.pub"
2020-01-02 21:19:59,628 INFO  [grid::actions::admin] Writing file: "/etc/grid/keys/gridd.priv"
```

**Comment by rbuysse** [comment]
It would be nice to have a flag to change the output directory as well. Keys will not always be generated on a machine with gridd installed.

File: cli/src/actions/admin.rs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    let public_key_path: PathBuf = [key_dir, "gridd.pub"].iter().collect();
+    let private_key_path: PathBuf = [key_dir, "gridd.priv"].iter().collect();
+
+    match conflict_strategy {
+        ConflictStrategy::Force => (),
+        ConflictStrategy::Skip => {
+            if public_key_path.exists() && !private_key_path.exists() {
+                return Err(CliError::UserError(format!(
+                    "{} already exists without a corresponding private key. \
+                     Rerun with --force to overwrite existing files",
```
Comment (chenette) [inline cli/src/actions/admin.rs]: Need a period at the end.

File: cli/src/actions/admin.rs [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+                    public_key_path.as_path().display(),
+                )));
+            }
+
+            if !public_key_path.exists() && private_key_path.exists() {
+                return Err(CliError::UserError(format!(
+                    "{} already exists without a corresponding public key. \
+                     Rerun with --force to overwrite existing files",
```
Comment (chenette) [inline cli/src/actions/admin.rs]: End with a period.

File: cli/src/actions/admin.rs [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+            if public_key_path.exists() && private_key_path.exists() {
+                println!("Admin private keys exists; skipping generation.");
+                return Ok(());
+            }
+        }
+        ConflictStrategy::Error => {
+            if public_key_path.exists() || private_key_path.exists() {
+                return Err(CliError::UserError(format!(
+                    "Key files already exist at {}. Rerun with --force to \
+                     overwrite existing files",
```
Comment (chenette) [inline cli/src/actions/admin.rs]: End with a period.

File: cli/src/actions/admin.rs [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+            if !public_key_path.exists() && private_key_path.exists() {
+                return Err(CliError::UserError(format!(
+                    "{} already exists without a corresponding public key. \
+                     Rerun with --force to overwrite existing files",
+                    private_key_path.as_path().display(),
+                )));
+            }
+
+            if public_key_path.exists() && private_key_path.exists() {
+                println!("Admin private keys exists; skipping generation.");
```
Comment (chenette) [inline cli/src/actions/admin.rs]: "key" (not "keys") - unless it should be "private keys exist".

File: cli/src/main.rs [hunk 8]
```diff
@@ -158,8 +162,46 @@ fn run() -> Result<(), CliError> {
...
+    {
+        use clap::{Arg, SubCommand};
+
+        app = app.subcommand(
+            SubCommand::with_name("admin")
+                .about("Administrative commands for gridd")
+                .setting(clap::AppSettings::SubcommandRequiredElseHelp)
+                .subcommand(
+                    SubCommand::with_name("keygen")
+                        .about("Generates keys with which gridd can sign transactions and batches.")
```
Comment (chenette) [inline cli/src/main.rs]: This is very minor, but I recommend "Generates keys for gridd to sign ...."

File: cli/src/actions/admin.rs (hunk 3 above)
Comment (peterschwarz) [reply cli/src/actions/admin.rs]: Fixed

File: cli/src/main.rs [hunk 9]
```diff
@@ -158,8 +162,46 @@ fn run() -> Result<(), CliError> {
...
+                            Arg::with_name("force")
+                                .long("force")
+                                .conflicts_with("skip")
+                                .help("Overwrite files if they exist"),
+                        )
+                        .arg(
+                            Arg::with_name("skip")
+                                .long("skip")
+                                .conflicts_with("force")
+                                .help("Check if files exists, generate if missing"),
```
Comment (chenette) [inline cli/src/main.rs]: Subject-verb mismatch: either "if file exists" or "if files exist".

Also, change the comma to a semicolon.

File: cli/src/actions/admin.rs (hunk 7 above)
Comment (peterschwarz) [reply cli/src/actions/admin.rs]: This has already been fixed

File: cli/src/actions/admin.rs (hunk 4 above)
Comment (peterschwarz) [reply cli/src/actions/admin.rs]: All error messages in this file are now consistent with no period.

Comment (peterschwarz) [reply cli/src/actions/admin.rs]: All error messages in this file are now consistent with no period.

File: cli/src/actions/admin.rs (hunk 5 above)
Comment (peterschwarz) [reply cli/src/actions/admin.rs]: All error messages in this file are now consistent with no period.

File: cli/src/actions/admin.rs (hunk 6 above)
Comment (peterschwarz) [reply cli/src/actions/admin.rs]: All error messages in this file are now consistent with no period.

File: cli/src/main.rs (hunk 9 above)
Comment (peterschwarz) [reply cli/src/main.rs]: Fixed

File: cli/src/main.rs (hunk 8 above)
Comment (peterschwarz) [reply cli/src/main.rs]: Updated

**Pull request merged**